### ISSUE TITLE
Problems with config parameters execution timing and evaluation

### ISSUE THREAD

**Issue by Hammarn** [nextflow-io/nextflow#469]
We have a problem with execution timing and evaluation for our config. [This](https://github.com/ewels/NGI-RNAseq/blob/32c7ebe1faa3ee7b8b07f88111cd294eb686475e/conf/igenomes.config) kinda works bit always ends up being ./iGenomes/ and not overwritten [here](https://github.com/ewels/NGI-RNAseq/blob/32c7ebe1faa3ee7b8b07f88111cd294eb686475e/conf/uppmax.config#L89).
We tried having {} to delay the evaluation until runtime but that gave the following error:
```
ERROR ~ org.codehaus.groovy.runtime.GStringImpl cannot be cast to java.nio.file.FileSystem
```
I included trace for both the first and second situation; 
[nextflow.log](https://github.com/nextflow-io/nextflow/files/1349294/nextflow.log)

[nextflow_without_brackets.log](https://github.com/nextflow-io/nextflow/files/1349299/nextflow_without_brackets.log)

**Comment by pditommaso** [comment]
I'm a bit confused. If I'm understanding it's a configuration file issue. Which is exactly the line raising the problem and the different alternative you are trying (write below please).

**Comment by Hammarn** [comment]
See this commit [here](https://github.com/ewels/NGI-RNAseq/commit/32c7ebe1faa3ee7b8b07f88111cd294eb686475e).
With the brackets the the string error above is thrown.
BUt without them the path is set to `./iGenomes/ ` and never updated.

It's [these](https://github.com/ewels/NGI-RNAseq/blob/483a3bef2787af267052daadb8f2e1d108bd0cc8/conf/igenomes.config#L17) lines that are the culprit:

**Comment by pditommaso** [comment]
Curly brackets should not be used in the config file params because they define a closure which is *not* resolved. The right syntax is the following 

```
 params {
  // illumina iGenomes reference file paths on UPPMAX
  igenomes_base = './iGenomes/'
  genomes {
    'GRCh37' {
      bed12   = "${params.igenomes_base}/Homo_sapiens/Ensembl/GRCh37/Annotation/Genes/genes.bed" 
      fasta   = "${params.igenomes_base}/Homo_sapiens/Ensembl/GRCh37/Sequence/WholeGenomeFasta/genome.fa" 
      gtf     = "${params.igenomes_base}/Homo_sapiens/Ensembl/GRCh37/Annotation/Genes/genes.gtf" 
      star    = "${params.igenomes_base}/Homo_sapiens/Ensembl/GRCh37/Sequence/STARIndex/" 
    }
  
    ... 
 } 
}
```

The following snippet: 

```
println params.genomes.GRCh37
```

Prints the expected output: 

```
[bed12:./iGenomes//Homo_sapiens/Ensembl/GRCh37/Annotation/Genes/genes.bed, fasta:./iGenomes//Homo_sapiens/Ensembl/GRCh37/Sequence/WholeGenomeFasta/genome.fa, gtf:./iGenomes//Homo_sapiens/Ensembl/GRCh37/Annotation/Genes/genes.gtf, star:./iGenomes//Homo_sapiens/Ensembl/GRCh37/Sequence/STARIndex/]
```

If I overrides the `igenomes_base` params specifying it on the command line, eg 

```
nextflow -c igenomes.config run p.nf --igenomes_base /foo
```

I get base path is returned as expected 

```
[bed12:/foo/Homo_sapiens/Ensembl/GRCh37/Annotation/Genes/genes.bed, fasta:/foo/Homo_sapiens/Ensembl/GRCh37/Sequence/WholeGenomeFasta/genome.fa, gtf:/foo/Homo_sapiens/Ensembl/GRCh37/Annotation/Genes/genes.gtf, star:/foo/Homo_sapiens/Ensembl/GRCh37/Sequence/STARIndex/]
```

What is the problem?

**Comment by Hammarn** [comment]
The problem is that we want the igenomes_base to be overwritten by another config file and that does not seem to work. 
According to our [main config](https://github.com/ewels/NGI-RNAseq/blob/master/nextflow.config#L13) the igenomes profile should be overwritten by the Uppmax.config [here](https://github.com/ewels/NGI-RNAseq/blob/master/conf/uppmax.config#L89) 

Having to specify it on the command line seem to work which is good but we would like to be able to just use the uppmax.config file - it's what it's there for.

**Comment by pditommaso** [comment]
Please provide a minimal test case I can reproduce.

**Comment by Hammarn** [comment]
I made a toy pipeline to test this out:
https://github.com/Hammarn/Nextflow_test

If run with the standard profile then we get the case with the igenomes not being overwritten.
If using `-profile test` then the string error is thrown.

**Comment by pditommaso** [comment]
Well, I'm getting the following error with and without `-profile test` but it's not a config error 

```
ERROR ~ No signature of method: ch.qos.logback.classic.Logger.info() is applicable for argument types: (java.lang.Boolean) values: [false]
Possible solutions: info(java.lang.String), info(java.lang.String, [Ljava.lang.Object;), info(java.lang.String, java.lang.Object), info(java.lang.String, java.lang.Throwable), info(org.slf4j.Marker, java.lang.String), info1(java.lang.Object)
```

The problem is that `log.info` requires a `String`, instead you are passing a `Boolean`. 

A apart of that without any profile the params for `GRCh37` is the following: 

```
[bed12:./iGenomes//Homo_sapiens/Ensembl/GRCh37/Annotation/Genes/genes.bed, fasta:./iGenomes//Homo_sapiens/Ensembl/GRCh37/Sequence/WholeGenomeFasta/genome.fa, gtf:./iGenomes//Homo_sapiens/Ensembl/GRCh37/Annotation/Genes/genes.gtf, star:./iGenomes//Homo_sapiens/Ensembl/GRCh37/Sequence/STARIndex/]
```

I guess that you are expecting instead: 

```
[bed12:/sw/data/uppnex/igenomes//Homo_sapiens/Ensembl/GRCh37/Annotation/Genes/genes.bed, fasta:/sw/data/uppnex/igenomes//Homo_sapiens/Ensembl/GRCh37/Sequence/WholeGenomeFasta/genome.fa, gtf:/sw/data/uppnex/igenomes//Homo_sapiens/Ensembl/GRCh37/Annotation/Genes/genes.gtf, star:/sw/data/uppnex/igenomes//Homo_sapiens/Ensembl/GRCh37/Sequence/STARIndex/]
```

I think what is happening here is that [this definition](https://github.com/Hammarn/Nextflow_test/blob/master/conf/igenomes.config#L14) is the one used that the `params` in the `igenomes.config` are evaluated whatever is the current (and the future) value of `igenomes_base`. 

I'm even sure if this is a bug or feature. A +1 for a bug would be the argument that `params` could be assigned only once. But also in this case your code won't work, because you are changing the value *after* in the [level3.config](https://github.com/Hammarn/Nextflow_test/blob/master/nextflow.config#L18) file. 

However it is possible to override a parameter value on the command line, for example: 

```
$ nextflow run main.nf --igenomes_base xxx

N E X T F L O W  ~  version 0.25.7
Launching `main.nf` [furious_wiles] - revision: e560d494fd
[bed12:xxx/Homo_sapiens/Ensembl/GRCh37/Annotation/Genes/genes.bed, fasta:xxx/Homo_sapiens/Ensembl/GRCh37/Sequence/WholeGenomeFasta/genome.fa, gtf:xxx/Homo_sapiens/Ensembl/GRCh37/Annotation/Genes/genes.gtf, star:xxx/Homo_sapiens/Ensembl/GRCh37/Sequence/STARIndex/]
```

**Comment by Hammarn** [comment]
@pditommaso Pinging you in case you forgot about this 😉

**Comment by pditommaso** [comment]
Well, I didn't forget. I've replied! ;)

**Comment by pditommaso** [comment]
Hi Rickard, did my suggestion helped? have you find a solution or this is still open?

**Comment by Hammarn** [comment]
Thanks for your input @pditommaso. Your reply hadn't showed up for me when I pinged you last. I guess I should have refreshed my browser. 
Thanks to @senthil10 we were able to figure out a way to fix this;

- Move the initial definition of `params.igenomes_base` is moved outside of the `igenomes.config`.
- Then place `igenomes.config` last in the config execution order.
- Then place the config with the path that you want before `igenomes.config` i.e. ´uppmax.config´ in our [use case](https://github.com/SciLifeLab/NGI-RNAseq/pull/154)

I don't think this needs any change within Nextflow since we where able to solve our issue ourselves.
